### ISSUE TITLE
[FR] Support AMD GPU system metrics

### ISSUE THREAD

**Issue by kaixin96** [mlflow/mlflow#14214]
### Willingness to contribute

Yes. I would be willing to contribute this feature with guidance from the MLflow community.

### Proposal Summary

Current GPUMonitor relies on pynvml, which only supports Nvidia GPUs. Adding AMD GPU support would be a helpful plus

### Motivation

> #### What is the use case for this feature?

Monitor AMD GPU usage under system metrics

> #### Why is this use case valuable to support for MLflow users in general?

AMD GPUs are quite common now. PyTorch and many other packages already support AMD GPUs and ROCm. A lot of MLflow users would benefit from this feature.

> #### Why is this use case valuable to support for your project(s) or organization?

I'm using azure & azure-mlflow. This use case would be very helpful for MI300 cluster users.

> #### Why is it currently difficult to achieve this use case?

I don't see any particular difficulties to achieve this. Simply extending existing api to support amdsmi would work.

### What component(s) does this bug affect?

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/deployments`: MLflow Deployments client APIs, server, and third-party Deployments integrations
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

### What interface(s) does this bug affect?

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

### What language(s) does this bug affect?

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

### What integration(s) does this bug affect?

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

**Comment by harupy** [comment]
@kaixin96 Makes sense. Added `help wanted` label for a volunteer.

**Comment by mayijun** [comment]
@kaixin96 i had a workaround before by monkey patching the Mlflow GPU monitoring code, as my org use NVidia MIG enabled GPU, but current GPU monitor code is not compatible to Nvidia MIG. i also added torch cuda memory into system metrics. so if you want you can do any thing similar to me.

import logging
import sys
from mlflow.system_metrics.metrics.base_metrics_monitor import BaseMetricsMonitor
import torch

_logger = logging.getLogger(__name__)

try:
    import pynvml
except ImportError:
    pass

class CustomGPUMonitor(BaseMetricsMonitor):
    """Custom class for monitoring GPU stats, including MIG devices."""

    def __init__(self):
        if "pynvml" not in sys.modules:
            raise ImportError(
                "`pynvml` is not installed, to log GPU metrics please run `pip install pynvml` "
                "to install it."
            )
        try:
            pynvml.nvmlInit()
        except pynvml.NVMLError as e:
            raise RuntimeError(f"Failed to initialize NVML, skip logging GPU metrics: {e}")

        super().__init__()
        self.num_gpus = pynvml.nvmlDeviceGetCount()
        self._init_mig_devices()

    def _init_mig_devices(self):
        # Initialize MIG device handles
        self.mig_device_handles = []
        for i in range(self.num_gpus):
            device_handle = pynvml.nvmlDeviceGetHandleByIndex(i)
            for j in range(pynvml.nvmlDeviceGetMaxMigDeviceCount(device_handle)):
                try:
                    mig_device_handle = pynvml.nvmlDeviceGetMigDeviceHandleByIndex(device_handle, j)
                    self.mig_device_handles.append((i, j, mig_device_handle))
                except pynvml.NVMLError_NotFound:
                    break

    def collect_metrics(self):
        for device_index, mig_index, handle in self.mig_device_handles:
            try:
                memory = pynvml.nvmlDeviceGetMemoryInfo(handle)
                self._metrics[f"gpu_{device_index}_mig_{mig_index}_memory_usage_percentage"].append(
                    round(memory.used / memory.total * 100, 1)
                )
                self._metrics[f"gpu_{device_index}_mig_{mig_index}_memory_usage_megabytes"].append(
                    memory.used / 1e6
                )
                self._metrics[f"torch_allocated_memory_GB"].append(
                     round(torch.cuda.memory_allocated() / 1024 / 1024 / 1024, 3)
                )
                self._metrics[f"torch_max_memory_reserved_GB"].append(
                     round(torch.cuda.max_memory_reserved() / 1024 / 1024 / 1024, 3)
                )
            except pynvml.NVMLError as e:
                _logger.warning(f"Encountered error {e} when trying to collect GPU metrics for "
                                f"GPU {device_index}, MIG {mig_index}.")

    def aggregate_metrics(self):
        return {k: round(sum(v) / len(v), 1) for k, v in self._metrics.items()}
 
 
 import mlflow
mlflow.set_tracking_uri(os.environ.get('MLFLOW_TRACKING_URI'))

# Monkey-patch the original GPUMonitor class in MLflow with the CustomGPUMonitor
import mlflow.system_metrics.metrics.gpu_monitor as original_gpu_module
original_gpu_module.GPUMonitor = CustomGPUMonitor
mlflow.enable_system_metrics_logging()
<img width="950" alt="image" src="https://github.com/user-attachments/assets/650250ac-4452-4192-8de1-d505a2099862" />


i use monkey patch as the Mlflow GPUmonitor class or interface is currently not exposed to user, but i believe it should be exposed , as it's highly related to hardware platform. end user need to do customization according to it's hardware

**Comment by ffund** [comment]
I noticed that this is implemented in #12694

**Comment by rahuja23** [comment]
Hi @ffund  is this issue stilll active? I would love to help if needed 😄

**Comment by rahuja23** [comment]
Hi Guys!
Is this issue still open?

**Comment by B-Step62** [comment]
No, it has been implemented by the above PR and released in 2.21!🙂
